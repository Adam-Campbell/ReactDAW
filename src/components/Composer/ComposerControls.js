import React from 'react';
import PropTypes from 'prop-types';
import SelectInput from '../SelectInput';

const ComposerControls = props => (
    <div className="composer__controls-container">
        <SelectInput
            inputId="composer-cursor-select"
            label="Select tool:"
            value={props.cursorValue}
            handleChange={props.updateCursorValue}
            options={[
                { value: 'pointer', text: 'Pointer' },
                { value: 'pencil', text: 'Pencil' }
            ]} 
        />
        <SelectInput 
            inputId="composer-duration-select"
            label="Select section duration:"
            value={props.durationValue.toString()}
            handleChange={props.updateDurationValue}
            options={[
                { value: '1', text: '1' },
                { value: '2', text: '2' },
                { value: '3', text: '3' },
                { value: '4', text: '4' }
            ]} 
        />
    </div>
);

ComposerControls.propTypes = {
    cursorValue: PropTypes.string.isRequired,
    updateCursorValue: PropTypes.func.isRequired,
    durationValue: PropTypes.number.isRequired,
    updateDurationValue: PropTypes.func.isRequired
};

export default ComposerControls;