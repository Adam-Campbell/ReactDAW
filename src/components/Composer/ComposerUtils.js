import Tone from 'tone';
import { generateId } from '../../helpers';
import { getWholeBarsFromString } from '../../sharedUtils';


/**
 * Given a selection state, and and an app state, creates the data structure required for pasting later.
 * @param {array} currentSelectionState - an array of the ids of the currently selected sections
 * @param {object} allSections - the complete sections state from the store
 * @param {array} allChannels - the complete channels state from the store  
 * @return {object} - data structure that can be used later to paste the selection. 
 */
export const createCopiedSectionsDataStructure = (optionsObject) => {
    const { 
        currentSelectionState,
        allSections,
        allChannels 
    } = optionsObject;
    // Declare variable to keep track of lowest index encountered
    let lowestIndex = null;
    // Map over the selection state array and create an array of section objects, with the channelId being
    // replaced by that channels index position within the channels array
    const sectionObjects = currentSelectionState.map(sectionId => {
        const section = { ...allSections[sectionId] };
        const index = allChannels.findIndex(channel => channel.id === section.channelId);
        if (index < lowestIndex || lowestIndex === null) {
            lowestIndex = index;
        }
        delete section.channelId;
        section.channelIndex = index;
        return section;
    });
    // return the array of section objects, and the lowest index that was encountered. 
    return {
        sectionObjects,
        lowestIndex
    };
}


/**
 * Given an array of section objects, it compares the start property of each object and returns the 
 * earliest start found. 
 * @param {array} sectionObjects - an array of section objects. 
 * @returns {string} - the earliest start time found, in a BBS string format '0:0:0' 
 */
export const findEarliestStartTime = (sectionObjects) => {
    let earliestSectionStart = null;
    for (let section of sectionObjects) {
        if (earliestSectionStart === null) {
            earliestSectionStart = section.start;
        } else {
            let a = getWholeBarsFromString(earliestSectionStart);
            let b = getWholeBarsFromString(section.start);
            if (b < a) {
                earliestSectionStart = section.start;
            }
        }
    }
    return earliestSectionStart;
};

/**
 * Creates and returns an array containing all of the data for generating the grid lines on the composer canvas, 
 * such that the returned array can be mapped over in the render function to render the lines on the canvas.
 * @param {number} canvasHeight - the height of the composer canvas
 * @param {number} canvasWidth - the width of the composer canvas
 * @param {number} channelsArrayLength - the length of the array holding all of the channels in the app state
 * @returns {array} - the constructed array
 */
export const createGridLinesArray = (optionsObject) => {
    const { 
        canvasHeight,
        canvasWidth,
        channelsArrayLength
     } = optionsObject;
    let linesArray = [];
    // create vertical lines
    for (let i = 0; i <= 200;  i++) {
        linesArray.push([i*48 , 40, i*48, canvasHeight]);
    }
    // create horizontal lines
    let verticalBound = Math.max(4, channelsArrayLength);
    for (let j = 0; j <= verticalBound; j++) {
        linesArray.push([0, j*70+40, canvasWidth, j*70+40]);
    }
    return linesArray;
}

/**
 * Create an array of objects containing all of the data needed to render the visual representations of
 * the sections on the canvas, such that this array can be mapped over in the main render function to
 * render the necessary elements on the canvas.
 * @param {array} allChannels - the array holding all of the channels data in the apps state
 * @param {object} allSections - the object holding all of the sections data in the apps state
 * @returns {array} - the constructed array
 */
export const createSectionRectsArray = (optionsObject) => {
    const { 
        allChannels,
        allSections
    } = optionsObject;
    let arr = [];
    for (let i = 0; i < allChannels.length; i++) {
        let channel = allChannels[i];
        for (let sectionId of channel.sectionIds) {
            let section = allSections[sectionId];
            let rectObject = {
                x: getWholeBarsFromString(section.start) * 48,
                y: (i * 70) + 40,
                width: section.numberOfBars * 48,
                height: 70,
                color: channel.color,
                sectionId: section.id
            }
            arr.push(rectObject);
        }
    }
    return arr;
}

/**
 * Contains the logic for creating section objects, including verifying that the object is valid. If it is
 * then the object is returned, and if not then null is returned.
 * @param {number} x - the x position used to calculate the section info, comes from a mouse event. Should 
 * already be adjusted for any scrolling before being passed to this function.
 * @param {number} y - the y position used to calculate the section info, comes from a mouse event. Should 
 * already be adjusted for any scrolling before being passed to this function.
 * @param {array} allChannels - array of all of the channels currently in the app state.
 * @param {number} numberOfBars - the number of bars the section will last for.
 * @returns {object} - a section object, or null if the section was invalid and couldn't be created.
 */
export const createSectionObject = (optionsObject) => {
    const { 
        x, 
        y,
        allChannels,
        numberOfBars
    } = optionsObject;

    if (y > allChannels.length * 70 + 40) {
        return null;
    }

    const channelIndex = Math.floor((y-40)/70);
    const channelId = allChannels[channelIndex].id;
    const startAsNumber = Math.floor(x/48);
    const sectionStartAsBBS = `${startAsNumber}:0:0`;
    const newSectionObject = {
        id: generateId(),
        channelId,
        notes: [],
        start: sectionStartAsBBS,
        numberOfBars,
    };
    return newSectionObject;
};

