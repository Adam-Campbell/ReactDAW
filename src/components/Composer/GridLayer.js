import React from 'react';
import PropTypes from 'prop-types';
import { Layer, Rect, Line } from 'react-konva';
import { UIColors } from '../../constants';

const GridLayer = props => (
    <Layer ref={props.gridLayerRef}>
        <Rect 
            x={0}
            y={0}
            width={props.canvasWidth}
            height={props.canvasHeight}
            fill={UIColors.backgroundDark}
        />
        {
            props.gridLinesArray.map((linePoints, index) => (
                <Line
                    points={linePoints}
                    listening={false}
                    stroke={UIColors.backgroundMedium}
                    strokeWidth={2}
                    shadowColor={UIColors.backgroundMedium}
                    shadowBlur={4}
                    shadowOffsetX={0}
                    shadowOffsetY={0}
                    key={index}
                />
            ))
        }
    </Layer>
);

GridLayer.propTypes = {
    gridLayerRef: PropTypes.object.isRequired,
    canvasWidth: PropTypes.number.isRequired,
    canvasHeight: PropTypes.number.isRequired,
    gridLinesArray: PropTypes.arrayOf(PropTypes.array).isRequired
};

export default GridLayer;