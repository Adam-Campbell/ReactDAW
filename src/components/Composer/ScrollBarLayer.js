import React from 'react';
import PropTypes from 'prop-types';
import { Layer, Rect } from 'react-konva';
import { UIColors } from '../../constants';

const ScrollBarLayer = props => (
    <Layer>
        <Rect 
            x={props.stageWidth - 24}
            y={0}
            width={24}
            height={props.stageHeight}
            fill={UIColors.lightPurple}
            shadowColor={UIColors.lightPurple}
            shadowBlur={4}
            shadowOffsetX={0}
            shadowOffsetY={0}
        />
        <Rect 
            x={props.stageWidth - 19}
            y={10}
            width={14}
            height={100}
            fill={UIColors.pink}
            draggable={true}
            dragBoundFunc={(pos) => {
                const currY = pos.y;
                // total stage height minus height of horizontal scroll bar
                // minus height of scroll slider
                const highBound = props.stageHeight - 24 - 100;
                pos.y = Math.min(Math.max(props.scrollPadding, currY), highBound);
                pos.x = props.stageWidth - 19;
                return pos;
            }}
            onDragMove={props.verticalDragMove}
            onClick={props.handleScrollBarClickEvents}
            onMouseDown={props.handleScrollBarClickEvents}
        />
        <Rect 
            x={0}
            y={props.stageHeight - 24}
            width={props.stageWidth}
            height={24}
            fill={UIColors.lightPurple}
            shadowColor={UIColors.lightPurple}
            shadowBlur={4}
            shadowOffsetX={0}
            shadowOffsetY={0}
        />
        <Rect 
            x={10}
            y={props.stageHeight - 19}
            width={100}
            height={14}
            fill={UIColors.pink}
            draggable={true}
            dragBoundFunc={(pos) => {
                const currX = pos.x;
                const highBound = props.stageWidth - 24 - 100;
                pos.x = Math.min(Math.max(currX, props.scrollPadding), highBound);
                pos.y = props.stageHeight - 19;
                return pos;
            }}
            onDragMove={props.horizontalDragMove}
            onClick={props.handleScrollBarClickEvents}
            onMouseDown={props.handleScrollBarClickEvents}
        />
    </Layer>
);

ScrollBarLayer.propTypes = {
    stageHeight: PropTypes.number.isRequired,
    stageWidth: PropTypes.number.isRequired,
    scrollPadding: PropTypes.number.isRequired,
    verticalDragMove: PropTypes.func.isRequired,
    horizontalDragMove: PropTypes.func.isRequired,
    handleScrollBarClickEvents: PropTypes.func.isRequired
};

export default ScrollBarLayer;