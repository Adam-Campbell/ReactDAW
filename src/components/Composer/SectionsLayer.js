import React from 'react';
import PropTypes from 'prop-types';
import { Layer, Rect, Line } from 'react-konva';
import { UIColors } from '../../constants';

const SectionsLayer = props => (
    <Layer ref={props.sectionsLayerRef}>
        {
            props.sectionRectsArray.map((section, index) => (
                <Rect 
                    x={section.x}
                    y={section.y}
                    sectionId={section.sectionId}
                    height={section.height}
                    width={section.width}
                    fill={
                        props.currentlySelectedSections.includes(section.sectionId) ? 
                        UIColors.darkGrey :
                        section.color
                    }
                    stroke={UIColors.offWhite}
                    strokeWidth={2}
                    type={'section'}
                    key={index}
                    onClick={(e) => props.handleSectionClick(e, section.sectionId)}
                    onDblClick={(e) => props.handleSectionDoubleClick(e, section.sectionId)}
                />
            ))
        } 
    </Layer>
);

SectionsLayer.propTypes = {
    sectionsLayerRef: PropTypes.object.isRequired,
    sectionRectsArray: PropTypes.arrayOf(PropTypes.object).isRequired,
    currentlySelectedSections: PropTypes.arrayOf(PropTypes.string).isRequired,
    handleSectionClick: PropTypes.func.isRequired,
    handleSectionDoubleClick: PropTypes.func.isRequired
};

export default SectionsLayer;