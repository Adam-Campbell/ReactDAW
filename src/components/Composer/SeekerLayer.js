import React from 'react';
import PropTypes from 'prop-types';
import { Layer, Line } from 'react-konva';
import { UIColors } from '../../constants';

const SeekerLayer = props => (
    <Layer ref={props.seekerLayerRef}>
        <Line
            ref={props.seekerLineRef}
            points={[0, 0, 0, props.canvasHeight]}
            listening={false}
            stroke={UIColors.offWhite}
            strokeWidth={2}
        />
    </Layer>
);

SeekerLayer.propTypes = {
    seekerLayerRef: PropTypes.object.isRequired,
    seekerLineRef: PropTypes.object.isRequired,
    canvasHeight: PropTypes.number.isRequired
}

export default SeekerLayer;