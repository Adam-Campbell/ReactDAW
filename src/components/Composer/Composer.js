import React from 'react';
import PropTypes from 'prop-types';
import { Stage } from 'react-konva';
import TrackInfo from './TrackInfo';
import AddTrackMenu from '../AddTrackMenu';
import ComposerControls from './ComposerControls';
import GridLayer from './GridLayer';
import SectionsLayer from './SectionsLayer';
import TransportLayer from './TransportLayer';
import SeekerLayer from './SeekerLayer';
import ScrollBarLayer from './ScrollBarLayer';

const Composer = props => (
    <div 
        className="composer__container"
        tabIndex="0"
        onKeyDown={props.handleKeyDown}
        style={{outline: 'none'}}
    >
        <ComposerControls 
            cursorValue={props.cursorValue}
            updateCursorValue={props.updateCursorValue}
            durationValue={props.durationValue}
            updateDurationValue={props.updateDurationValue}
        />
        <div className="composer__row">
            <div className="composer__track-info-container-outer">
                <AddTrackMenu />
                <div className="composer__track-info-scroll-outer">
                    <div 
                        className="composer__track-info-scroll-inner"
                        style={{top: props.trackInfoMenuTopScroll}}
                    >
                        {
                            props.channels.map((channel, index) => (
                                <TrackInfo 
                                    trackId={channel.id}
                                    key={channel.id}
                                    isSelected={props.currentlySelectedChannel === channel.id}
                                    updateSelectedChannel={props.updateSelectedChannel}
                                /> 
                            ))
                        }
                    </div>
                </div>
            </div>
            <div className="composer__canvas-container" id="composer-canvas-container">
                <Stage
                    container={'composer-canvas-container'}
                    onClick={props.handleStageClick}
                    onMouseDown={props.handleStageMouseDown}
                    onMouseUp={props.handleStageMouseUp}
                    ref={props.stageRef}
                    width={740} 
                    height={300} 
                >
                    <GridLayer 
                        gridLayerRef={props.gridLayerRef}
                        canvasWidth={props.canvasWidth}
                        canvasHeight={props.canvasHeight}
                        gridLinesArray={props.gridLinesArray}
                    />
                    <SectionsLayer 
                        sectionsLayerRef={props.sectionsLayerRef}
                        sectionRectsArray={props.sectionRectsArray}
                        currentlySelectedSections={props.currentlySelectedSections}
                        handleSectionClick={props.handleSectionClick}
                        handleSectionDoubleClick={props.handleSectionDoubleClick}
                    />
                    <TransportLayer 
                        transportLayerRef={props.transportLayerRef}
                        canvasWidth={props.canvasWidth}
                    />
                    <SeekerLayer 
                        seekerLayerRef={props.seekerLayerRef}
                        seekerLineRef={props.seekerLineRef}
                        canvasHeight={props.canvasHeight}
                    />
                    <ScrollBarLayer 
                        stageHeight={props.stageHeight}
                        stageWidth={props.stageWidth}
                        scrollPadding={props.scrollPadding}
                        verticalDragMove={props.verticalDragMove}
                        horizontalDragMove={props.horizontalDragMove}
                        handleScrollBarClickEvents={props.handleScrollBarClickEvents}
                    />
                </Stage>
            </div>
        </div>
    </div>
);

Composer.propTypes = {
    // react refs
    stageRef: PropTypes.object.isRequired,
    gridLayerRef: PropTypes.object.isRequired,
    sectionsLayerRef: PropTypes.object.isRequired,
    transportLayerRef: PropTypes.object.isRequired,
    seekerLayerRef: PropTypes.object.isRequired,
    seekerLineRef: PropTypes.object.isRequired,
    // canvas figures
    canvasWidth: PropTypes.number.isRequired,
    canvasHeight: PropTypes.number.isRequired,
    stageHeight: PropTypes.number.isRequired,
    scrollPadding: PropTypes.number.isRequired,
    stageWidth: PropTypes.number.isRequired,
    // constructed arrays
    gridLinesArray: PropTypes.arrayOf(PropTypes.array).isRequired,
    sectionRectsArray: PropTypes.arrayOf(PropTypes.object).isRequired,
    // input values
    cursorValue: PropTypes.string.isRequired,
    durationValue: PropTypes.number.isRequired,
    // callback functions
    handleKeyDown: PropTypes.func.isRequired,
    updateCursorValue: PropTypes.func.isRequired,
    updateDurationValue: PropTypes.func.isRequired,
    handleStageClick: PropTypes.func.isRequired,
    handleStageMouseDown: PropTypes.func.isRequired,
    handleStageMouseUp: PropTypes.func.isRequired,
    handleSectionClick: PropTypes.func.isRequired,
    handleSectionDoubleClick: PropTypes.func.isRequired,
    verticalDragMove: PropTypes.func.isRequired,
    horizontalDragMove: PropTypes.func.isRequired,
    updateSelectedChannel: PropTypes.func.isRequired,
    handleScrollBarClickEvents: PropTypes.func.isRequired,
    // other
    trackInfoMenuTopScroll: PropTypes.number.isRequired,
    channels: PropTypes.arrayOf(PropTypes.object).isRequired,
    currentlySelectedSections: PropTypes.arrayOf(PropTypes.string).isRequired,
    currentlySelectedChannel: PropTypes.any
};

export default Composer;