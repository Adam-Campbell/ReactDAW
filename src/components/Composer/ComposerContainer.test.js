import React from 'react';
import { ComposerContainer } from './';
import Composer from './Composer';
import { shallow } from 'enzyme';

const channelsState = [
    {
      id: '6984087700822156',
      name: 'Channel 1',
      color: '#d86597',
      instrumentId: '0374798538126734',
      effectIds: [],
      sectionIds: [
        '7009247242625234'
      ]
    }
];

const sectionsState = {
    '7009247242625234': {
      id: '7009247242625234',
      channelId: '6984087700822156',
      notes: [],
      start: '0:0:0',
      numberOfBars: 4
    }
};

test('renders correctly', () => {
    document.querySelector = jest.fn(() => ({
        getBoundingClientRect: () => ({
            height: 50
        })
    }));
    const component = shallow(
        <ComposerContainer 
            channels={channelsState}
            sections={sectionsState}
            isPlaying={false}
            addChannel={jest.fn()}
            removeChannel={jest.fn()}
            addSection={jest.fn()}
            removeSection={jest.fn()}
            openWindow={jest.fn()}
        />
    );
    expect(component).toMatchSnapshot();
    expect(component.find(Composer)).toHaveLength(1);
    expect(component.find(Composer).first().props().channels).toEqual(channelsState);
});
