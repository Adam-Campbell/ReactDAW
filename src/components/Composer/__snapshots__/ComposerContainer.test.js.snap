// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Composer
  canvasHeight={320}
  canvasWidth={9600}
  channels={
    Array [
      Object {
        "color": "#d86597",
        "effectIds": Array [],
        "id": "6984087700822156",
        "instrumentId": "0374798538126734",
        "name": "Channel 1",
        "sectionIds": Array [
          "7009247242625234",
        ],
      },
    ]
  }
  currentlySelectedChannel={null}
  currentlySelectedSections={Array []}
  cursorValue="pointer"
  durationValue={4}
  gridLayerRef={
    Object {
      "current": null,
    }
  }
  gridLinesArray={
    Array [
      Array [
        0,
        40,
        0,
        320,
      ],
      Array [
        48,
        40,
        48,
        320,
      ],
      Array [
        96,
        40,
        96,
        320,
      ],
      Array [
        144,
        40,
        144,
        320,
      ],
      Array [
        192,
        40,
        192,
        320,
      ],
      Array [
        240,
        40,
        240,
        320,
      ],
      Array [
        288,
        40,
        288,
        320,
      ],
      Array [
        336,
        40,
        336,
        320,
      ],
      Array [
        384,
        40,
        384,
        320,
      ],
      Array [
        432,
        40,
        432,
        320,
      ],
      Array [
        480,
        40,
        480,
        320,
      ],
      Array [
        528,
        40,
        528,
        320,
      ],
      Array [
        576,
        40,
        576,
        320,
      ],
      Array [
        624,
        40,
        624,
        320,
      ],
      Array [
        672,
        40,
        672,
        320,
      ],
      Array [
        720,
        40,
        720,
        320,
      ],
      Array [
        768,
        40,
        768,
        320,
      ],
      Array [
        816,
        40,
        816,
        320,
      ],
      Array [
        864,
        40,
        864,
        320,
      ],
      Array [
        912,
        40,
        912,
        320,
      ],
      Array [
        960,
        40,
        960,
        320,
      ],
      Array [
        1008,
        40,
        1008,
        320,
      ],
      Array [
        1056,
        40,
        1056,
        320,
      ],
      Array [
        1104,
        40,
        1104,
        320,
      ],
      Array [
        1152,
        40,
        1152,
        320,
      ],
      Array [
        1200,
        40,
        1200,
        320,
      ],
      Array [
        1248,
        40,
        1248,
        320,
      ],
      Array [
        1296,
        40,
        1296,
        320,
      ],
      Array [
        1344,
        40,
        1344,
        320,
      ],
      Array [
        1392,
        40,
        1392,
        320,
      ],
      Array [
        1440,
        40,
        1440,
        320,
      ],
      Array [
        1488,
        40,
        1488,
        320,
      ],
      Array [
        1536,
        40,
        1536,
        320,
      ],
      Array [
        1584,
        40,
        1584,
        320,
      ],
      Array [
        1632,
        40,
        1632,
        320,
      ],
      Array [
        1680,
        40,
        1680,
        320,
      ],
      Array [
        1728,
        40,
        1728,
        320,
      ],
      Array [
        1776,
        40,
        1776,
        320,
      ],
      Array [
        1824,
        40,
        1824,
        320,
      ],
      Array [
        1872,
        40,
        1872,
        320,
      ],
      Array [
        1920,
        40,
        1920,
        320,
      ],
      Array [
        1968,
        40,
        1968,
        320,
      ],
      Array [
        2016,
        40,
        2016,
        320,
      ],
      Array [
        2064,
        40,
        2064,
        320,
      ],
      Array [
        2112,
        40,
        2112,
        320,
      ],
      Array [
        2160,
        40,
        2160,
        320,
      ],
      Array [
        2208,
        40,
        2208,
        320,
      ],
      Array [
        2256,
        40,
        2256,
        320,
      ],
      Array [
        2304,
        40,
        2304,
        320,
      ],
      Array [
        2352,
        40,
        2352,
        320,
      ],
      Array [
        2400,
        40,
        2400,
        320,
      ],
      Array [
        2448,
        40,
        2448,
        320,
      ],
      Array [
        2496,
        40,
        2496,
        320,
      ],
      Array [
        2544,
        40,
        2544,
        320,
      ],
      Array [
        2592,
        40,
        2592,
        320,
      ],
      Array [
        2640,
        40,
        2640,
        320,
      ],
      Array [
        2688,
        40,
        2688,
        320,
      ],
      Array [
        2736,
        40,
        2736,
        320,
      ],
      Array [
        2784,
        40,
        2784,
        320,
      ],
      Array [
        2832,
        40,
        2832,
        320,
      ],
      Array [
        2880,
        40,
        2880,
        320,
      ],
      Array [
        2928,
        40,
        2928,
        320,
      ],
      Array [
        2976,
        40,
        2976,
        320,
      ],
      Array [
        3024,
        40,
        3024,
        320,
      ],
      Array [
        3072,
        40,
        3072,
        320,
      ],
      Array [
        3120,
        40,
        3120,
        320,
      ],
      Array [
        3168,
        40,
        3168,
        320,
      ],
      Array [
        3216,
        40,
        3216,
        320,
      ],
      Array [
        3264,
        40,
        3264,
        320,
      ],
      Array [
        3312,
        40,
        3312,
        320,
      ],
      Array [
        3360,
        40,
        3360,
        320,
      ],
      Array [
        3408,
        40,
        3408,
        320,
      ],
      Array [
        3456,
        40,
        3456,
        320,
      ],
      Array [
        3504,
        40,
        3504,
        320,
      ],
      Array [
        3552,
        40,
        3552,
        320,
      ],
      Array [
        3600,
        40,
        3600,
        320,
      ],
      Array [
        3648,
        40,
        3648,
        320,
      ],
      Array [
        3696,
        40,
        3696,
        320,
      ],
      Array [
        3744,
        40,
        3744,
        320,
      ],
      Array [
        3792,
        40,
        3792,
        320,
      ],
      Array [
        3840,
        40,
        3840,
        320,
      ],
      Array [
        3888,
        40,
        3888,
        320,
      ],
      Array [
        3936,
        40,
        3936,
        320,
      ],
      Array [
        3984,
        40,
        3984,
        320,
      ],
      Array [
        4032,
        40,
        4032,
        320,
      ],
      Array [
        4080,
        40,
        4080,
        320,
      ],
      Array [
        4128,
        40,
        4128,
        320,
      ],
      Array [
        4176,
        40,
        4176,
        320,
      ],
      Array [
        4224,
        40,
        4224,
        320,
      ],
      Array [
        4272,
        40,
        4272,
        320,
      ],
      Array [
        4320,
        40,
        4320,
        320,
      ],
      Array [
        4368,
        40,
        4368,
        320,
      ],
      Array [
        4416,
        40,
        4416,
        320,
      ],
      Array [
        4464,
        40,
        4464,
        320,
      ],
      Array [
        4512,
        40,
        4512,
        320,
      ],
      Array [
        4560,
        40,
        4560,
        320,
      ],
      Array [
        4608,
        40,
        4608,
        320,
      ],
      Array [
        4656,
        40,
        4656,
        320,
      ],
      Array [
        4704,
        40,
        4704,
        320,
      ],
      Array [
        4752,
        40,
        4752,
        320,
      ],
      Array [
        4800,
        40,
        4800,
        320,
      ],
      Array [
        4848,
        40,
        4848,
        320,
      ],
      Array [
        4896,
        40,
        4896,
        320,
      ],
      Array [
        4944,
        40,
        4944,
        320,
      ],
      Array [
        4992,
        40,
        4992,
        320,
      ],
      Array [
        5040,
        40,
        5040,
        320,
      ],
      Array [
        5088,
        40,
        5088,
        320,
      ],
      Array [
        5136,
        40,
        5136,
        320,
      ],
      Array [
        5184,
        40,
        5184,
        320,
      ],
      Array [
        5232,
        40,
        5232,
        320,
      ],
      Array [
        5280,
        40,
        5280,
        320,
      ],
      Array [
        5328,
        40,
        5328,
        320,
      ],
      Array [
        5376,
        40,
        5376,
        320,
      ],
      Array [
        5424,
        40,
        5424,
        320,
      ],
      Array [
        5472,
        40,
        5472,
        320,
      ],
      Array [
        5520,
        40,
        5520,
        320,
      ],
      Array [
        5568,
        40,
        5568,
        320,
      ],
      Array [
        5616,
        40,
        5616,
        320,
      ],
      Array [
        5664,
        40,
        5664,
        320,
      ],
      Array [
        5712,
        40,
        5712,
        320,
      ],
      Array [
        5760,
        40,
        5760,
        320,
      ],
      Array [
        5808,
        40,
        5808,
        320,
      ],
      Array [
        5856,
        40,
        5856,
        320,
      ],
      Array [
        5904,
        40,
        5904,
        320,
      ],
      Array [
        5952,
        40,
        5952,
        320,
      ],
      Array [
        6000,
        40,
        6000,
        320,
      ],
      Array [
        6048,
        40,
        6048,
        320,
      ],
      Array [
        6096,
        40,
        6096,
        320,
      ],
      Array [
        6144,
        40,
        6144,
        320,
      ],
      Array [
        6192,
        40,
        6192,
        320,
      ],
      Array [
        6240,
        40,
        6240,
        320,
      ],
      Array [
        6288,
        40,
        6288,
        320,
      ],
      Array [
        6336,
        40,
        6336,
        320,
      ],
      Array [
        6384,
        40,
        6384,
        320,
      ],
      Array [
        6432,
        40,
        6432,
        320,
      ],
      Array [
        6480,
        40,
        6480,
        320,
      ],
      Array [
        6528,
        40,
        6528,
        320,
      ],
      Array [
        6576,
        40,
        6576,
        320,
      ],
      Array [
        6624,
        40,
        6624,
        320,
      ],
      Array [
        6672,
        40,
        6672,
        320,
      ],
      Array [
        6720,
        40,
        6720,
        320,
      ],
      Array [
        6768,
        40,
        6768,
        320,
      ],
      Array [
        6816,
        40,
        6816,
        320,
      ],
      Array [
        6864,
        40,
        6864,
        320,
      ],
      Array [
        6912,
        40,
        6912,
        320,
      ],
      Array [
        6960,
        40,
        6960,
        320,
      ],
      Array [
        7008,
        40,
        7008,
        320,
      ],
      Array [
        7056,
        40,
        7056,
        320,
      ],
      Array [
        7104,
        40,
        7104,
        320,
      ],
      Array [
        7152,
        40,
        7152,
        320,
      ],
      Array [
        7200,
        40,
        7200,
        320,
      ],
      Array [
        7248,
        40,
        7248,
        320,
      ],
      Array [
        7296,
        40,
        7296,
        320,
      ],
      Array [
        7344,
        40,
        7344,
        320,
      ],
      Array [
        7392,
        40,
        7392,
        320,
      ],
      Array [
        7440,
        40,
        7440,
        320,
      ],
      Array [
        7488,
        40,
        7488,
        320,
      ],
      Array [
        7536,
        40,
        7536,
        320,
      ],
      Array [
        7584,
        40,
        7584,
        320,
      ],
      Array [
        7632,
        40,
        7632,
        320,
      ],
      Array [
        7680,
        40,
        7680,
        320,
      ],
      Array [
        7728,
        40,
        7728,
        320,
      ],
      Array [
        7776,
        40,
        7776,
        320,
      ],
      Array [
        7824,
        40,
        7824,
        320,
      ],
      Array [
        7872,
        40,
        7872,
        320,
      ],
      Array [
        7920,
        40,
        7920,
        320,
      ],
      Array [
        7968,
        40,
        7968,
        320,
      ],
      Array [
        8016,
        40,
        8016,
        320,
      ],
      Array [
        8064,
        40,
        8064,
        320,
      ],
      Array [
        8112,
        40,
        8112,
        320,
      ],
      Array [
        8160,
        40,
        8160,
        320,
      ],
      Array [
        8208,
        40,
        8208,
        320,
      ],
      Array [
        8256,
        40,
        8256,
        320,
      ],
      Array [
        8304,
        40,
        8304,
        320,
      ],
      Array [
        8352,
        40,
        8352,
        320,
      ],
      Array [
        8400,
        40,
        8400,
        320,
      ],
      Array [
        8448,
        40,
        8448,
        320,
      ],
      Array [
        8496,
        40,
        8496,
        320,
      ],
      Array [
        8544,
        40,
        8544,
        320,
      ],
      Array [
        8592,
        40,
        8592,
        320,
      ],
      Array [
        8640,
        40,
        8640,
        320,
      ],
      Array [
        8688,
        40,
        8688,
        320,
      ],
      Array [
        8736,
        40,
        8736,
        320,
      ],
      Array [
        8784,
        40,
        8784,
        320,
      ],
      Array [
        8832,
        40,
        8832,
        320,
      ],
      Array [
        8880,
        40,
        8880,
        320,
      ],
      Array [
        8928,
        40,
        8928,
        320,
      ],
      Array [
        8976,
        40,
        8976,
        320,
      ],
      Array [
        9024,
        40,
        9024,
        320,
      ],
      Array [
        9072,
        40,
        9072,
        320,
      ],
      Array [
        9120,
        40,
        9120,
        320,
      ],
      Array [
        9168,
        40,
        9168,
        320,
      ],
      Array [
        9216,
        40,
        9216,
        320,
      ],
      Array [
        9264,
        40,
        9264,
        320,
      ],
      Array [
        9312,
        40,
        9312,
        320,
      ],
      Array [
        9360,
        40,
        9360,
        320,
      ],
      Array [
        9408,
        40,
        9408,
        320,
      ],
      Array [
        9456,
        40,
        9456,
        320,
      ],
      Array [
        9504,
        40,
        9504,
        320,
      ],
      Array [
        9552,
        40,
        9552,
        320,
      ],
      Array [
        9600,
        40,
        9600,
        320,
      ],
      Array [
        0,
        40,
        9600,
        40,
      ],
      Array [
        0,
        110,
        9600,
        110,
      ],
      Array [
        0,
        180,
        9600,
        180,
      ],
      Array [
        0,
        250,
        9600,
        250,
      ],
      Array [
        0,
        320,
        9600,
        320,
      ],
    ]
  }
  handleKeyDown={[Function]}
  handleScrollBarClickEvents={[Function]}
  handleSectionClick={[Function]}
  handleSectionDoubleClick={[Function]}
  handleStageClick={[Function]}
  handleStageMouseDown={[Function]}
  handleStageMouseUp={[Function]}
  horizontalDragMove={[Function]}
  scrollPadding={10}
  sectionRectsArray={
    Array [
      Object {
        "color": "#d86597",
        "height": 70,
        "sectionId": "7009247242625234",
        "width": 192,
        "x": 0,
        "y": 40,
      },
    ]
  }
  sectionsLayerRef={
    Object {
      "current": null,
    }
  }
  seekerLayerRef={
    Object {
      "current": null,
    }
  }
  seekerLineRef={
    Object {
      "current": null,
    }
  }
  stageHeight={300}
  stageRef={
    Object {
      "current": null,
    }
  }
  stageWidth={740}
  trackInfoMenuTopScroll={0}
  transportLayerRef={
    Object {
      "current": null,
    }
  }
  updateCursorValue={[Function]}
  updateDurationValue={[Function]}
  updateSelectedChannel={[Function]}
  verticalDragMove={[Function]}
/>
`;
